### PR TITLE
✨ Add new component intro table

### PR DESCRIPTION
**Pull request by robinvanopstal**
This adds all the needed functionality to show the new automatically generated component intro tables by using the data provided by [@ampproject/amp-toolbox/validator-rules](https://github.com/ampproject/amp-toolbox/tree/master/packages/validator-rules).

### Before
Bildschirmfoto 2020-01-06 um 21 29 59

### After
Bildschirmfoto 2020-01-06 um 21 29 51

Additionally this refactors the importer to be around 20x faster than before and moves all document related logic to a separate class.

Open tasks: design review with Longo, checking why the GrowReferenceChecker has problems with the new files, removing the manually maintained tables from the documents over at ampproject/amphtml and add missing teaser where needed.

### CODE REVIEW

**Comment by matthiasrohmer** [comment]
@googlebot I consent.

File: platform/lib/build/samplesBuilder.js [hunk 1]
```diff
@@ -570,6 +584,23 @@ class SamplesBuilder {
       }
     });
 
+    // Store the sample by it's used component to show all samples for a specific
```
Comment (tharders) [inline platform/lib/build/samplesBuilder.js]: A method that is named get... should never modify the state.
I know the code snipped above did already modify the parsedSample which is also bad.
I think we should refactor this.
Maybe call some methods in _createDocumentation that do the changes and then call getTeaserData with the usedComponents.

File: platform/lib/pipeline/componentReferenceDocument.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+const EXTENSION_TYPE_ELEMENT = 'element';
+const EXTENSION_TYPE_TEMPLATE = 'template';
+
+
+class ComponentReferenceDocument extends MarkdownDocument {
+  constructor(path, contents, extension) {
+    super(path, contents);
+
+    if (!this.teaser.text) {
+      throw new Error('Teaser missing in', extension.name);
```
Comment (tharders) [inline platform/lib/pipeline/componentReferenceDocument.js]: Was it discussed that documents without a teaser text should be skipped completely?

File: platform/lib/pipeline/gitHubImporter.js [hunk 3]
```diff
@@ -63,6 +63,21 @@ class GitHubImporter {
     return this.fetchContents_(filePath, repo, master);
   }
 
+  /**
+   * Fetches all files contained in a directory flattened down to
```
Comment (tharders) [inline platform/lib/pipeline/gitHubImporter.js]: The comment suggests the file contens are returned rather than the file path.

File: frontend/templates/views/partials/component-intro.j2 [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    <div class="ap-o-component-intro-box-scripts-item">
+      <code>{{ script }}</code>
+      <amp-iframe
+          width="24"
+          height="24"
+          layout="fixed"
+          sandbox="allow-scripts"
+          src="{{ podspec.base_urls.platform }}/shared/copy-script#{{ script|urlencode }}">
+          <div placeholder>
+            ..
```
Comment (sebastianbenz) [inline frontend/templates/views/partials/component-intro.j2]: a better way would be to use a disabled version of the copy icon as placeholder

File: frontend/templates/views/partials/component-intro.j2 [hunk 5]
```diff
@@ -0,0 +1,61 @@
+{% do doc.styles.addCssFile('css/components/molecules/code-snippet.css') %}
+{% do doc.amp_dependencies.add('amp-fx-collection', '0.1') %}
+
+<div class="ap-o-component-intro">
+  <h2 class="ap-o-component-intro-headline">Description</h2>
```
Comment (sebastianbenz) [inline frontend/templates/views/partials/component-intro.j2]: Not sure we need the header "Description" here

File: pages/content/shared/copy-script.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    html, body { background: none transparent; }
+    body { display: flex; min-height: 100vh; align-items: center; }
+    svg { fill: #005af0; }
+    input { opacity: 0 }
+  </style>
+</head>
+<body>
+  <svg style="position: absolute; width: 0; height: 0; overflow: hidden;" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <defs>
+      {{ doc.icons.emit() }}
```
Comment (sebastianbenz) [inline pages/content/shared/copy-script.html]: Are these all icons or only the copy one?

File: pages/content/shared/copy-script.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    </defs>
+  </svg>
+  <div class="ap-a-ico">
+    {% do doc.icons.useIcon('/icons/clipboard.svg') %}
+    <svg><use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="#clipboard"></use></svg>
+  </div>
+  <input id="script-tag">
+  <script>
+    const $scriptTag = document.getElementById('script-tag');
```
Comment (sebastianbenz) [inline pages/content/shared/copy-script.html]: don't use `$` as prefix

File: pages/content/shared/copy-script.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    const scriptTag = decodeURIComponent(window.location.hash.substr(1));
```
Comment (sebastianbenz) [inline pages/content/shared/copy-script.html]: rename to `stringToCopy`

File: platform/lib/build/samplesBuilder.js [hunk 9]
```diff
@@ -558,7 +572,7 @@ class SamplesBuilder {
    * @param  {Object} parsedSample
    * @return {Object}
    */
-  _getUsedComponents(parsedSample) {
+  _getUsedComponents(sample, parsedSample) {
```
Comment (sebastianbenz) [inline platform/lib/build/samplesBuilder.js]: please extract this into a separate file as this might be useful in other places as well.

File: platform/lib/pipeline/componentReferenceImporter.js [hunk 10]
```diff
@@ -15,289 +15,220 @@
  */
 require('module-alias/register');
 
-// By default, reference docs are imported from the latest release tag.
-// If a doc is broken in a release, add it to this this list to fetch from master instead.
-//
-// DON'T FORGET TO REMOVE ONCE IT'S FIXED
-const DOCS_TO_FETCH_FROM_MASTER = ['amp-script', 'amp-carousel'];
```
Comment (sebastianbenz) [inline platform/lib/pipeline/componentReferenceImporter.js]: Where did this go?

**Comment by sebastianbenz** [comment]
Noticed a few more things:

* the copy button should use `cursor: pointer;` 
* vertical pre padding not consistent for scripts
Screen Shot 2020-01-08 at 13 48 45

* layouts are not aligned with heading

Screen Shot 2020-01-08 at 13 48 41
* Copy icon cut off on mobile 
Screen Shot 2020-01-08 at 13 52 00
* two column layout does not work on mobile
Screen Shot 2020-01-08 at 13 52 33
* layout page does not work

```

[TIMING] localhost:8080/documentation/guides-and-tutorials/learn/amp-html-layout/layouts_demonstrated/?format=websites 0.086s [200]
500: /documentation/guides-and-tutorials/learn/amp-html-layout/layouts_demonstrated.html - 'grow.documents.document.Document object' has no attribute 'teaser'
Traceback (most recent call last):
  File "/Users/sbenz/Library/Python/2.7/lib/python/site-packages/grow/server/main.py", line 28, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/sbenz/Library/Python/2.7/lib/python/site-packages/grow/server/main.py", line 70, in wsgi_app
    response = self.dispatch_request(request)
  File "/Users/sbenz/Library/Python/2.7/lib/python/site-packages/grow/server/main.py", line 66, in dispatch_request
    return handlers.serve_pod(self.pod, request, matched)
  File "/Users/sbenz/Library/Python/2.7/lib/python/site-packages/grow/server/handlers.py", line 108, in serve_pod
    rendered_document = controller.render(jinja_env=jinja_env, request=request)
  File "/Users/sbenz/Library/Python/2.7/lib/python/site-packages/grow/rendering/render_controller.py", line 246, in render
    raise exception
BuildError: 'grow.documents.document.Document object' has no attribute 'teaser'
```
* The example list should always show the component specific example first (e.g. `amp-analytics`). Can we find a better name for the component specific samples? E.g.: `Component samples: amp-analytics`

File: pages/content/shared/copy-script.html (hunk 7 above)
Comment (matthiasrohmer) [reply pages/content/shared/copy-script.html]: Just curious: why? I'm used to prefixing variables containing DOM elements with $ to know that the stored object has DOM API methods.

**Comment by sebastianbenz** [comment]
What's the status here?

**Comment by LongoOfficial** [comment]
@sebastianbenz There was one issue on smaller screens where the layout draws too many edges on a small space. We would suggest to adjust the padding of the intro background on small screen sizes so it goes from edge to egde. This will solve the visual nesting problem without breaking the whole design system. See attached screenshot. wdyt?
localhost_8080_documentation_components_amp-carousel__format=websites(iPhone X) (1)

**Comment by sebastianbenz** [comment]
@LongoOfficial can you share a full mock that includes normal text as well?

**Comment by LongoOfficial** [comment]
@sebastianbenz 
localhost_8080_documentation_components_amp-carousel__format=websites(iPhone X) (1)

**Comment by LongoOfficial** [comment]
oh! and for the Component Version UI in the title, we imagined something like this:

50157f2e-38e1-4f1a-b1c8-397272f3e41a
70f8735b-e02f-485a-94d1-eae8d72b9d64

**Comment by sebastianbenz** [comment]
+1 to the version UI

re margins: can we half these on mobile? I still think they're too big (for normal text). This probably means that we have to move the main content onto the background, but would also help with our problem of having too many layers.

**Comment by LongoOfficial** [comment]
let's discuss it separately, as this will affect more than the component intro. I'll join the status call on monday.

File: platform/lib/pipeline/componentReferenceDocument.js (hunk 2 above)
Comment (matthiasrohmer) [reply platform/lib/pipeline/componentReferenceDocument.js]: As before a teaser text is now automatically extracted.

**Comment by matthiasrohmer** [comment]
@sebastianbenz, to quickly give you an update: there were some problems with `MarkdownDocument` implicitly altering `MarkdownDocument._contents` which are now resolved and make things a lot easier to grasp. While doing so I also introduced some more tests for all the rewriting stuff to be on the safe side. The new version toggle is also implemented.

A problem that still persists is that sometimes downloading files randomly fails with:

```
[6:51:27 PM] [ComponentReferenceImporter] › ✖  error     Failed to fetch extensions/amp-web-push/amp-web-push.md { Error: getaddrinfo ENOTFOUND api.github.com api.github.com:443
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:56:26)
  cause:
   { Error: getaddrinfo ENOTFOUND api.github.com api.github.com:443
       at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:56:26)
     errno: 'ENOTFOUND',
     code: 'ENOTFOUND',
     syscall: 'getaddrinfo',
     hostname: 'api.github.com',
     host: 'api.github.com',
     port: 443 },
  isOperational: true,
  errno: 'ENOTFOUND',
  code: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'api.github.com',
  host: 'api.github.com',
  port: 443 }
```

This could be due to rate limiting which I'll check tomorrow. 

For all comments related to the `SamplesBuilder` I'll create a separate issue as it needs some refactoring anyway.

The plan is to create a PR towards the ampproject/amphtml repo tomorrow and finalise the last bits that are missing in here to then finally have this going 🙂

**Comment by matthiasrohmer** [comment]
Final testing in progress, related PR issued: https://github.com/ampproject/amphtml/pull/26383.

**Comment by matthiasrohmer** [comment]
Tests went well, comments in the related PR are taken care of today. I'm going to reimplement the Pull-from-master logic today and also stumbled over https://github.com/ampproject/amp.dev/issues/3023 which I want to double check.

File: pages/content/shared/copy-script.html (hunk 6 above)
Comment (matthiasrohmer) [reply pages/content/shared/copy-script.html]: This will only emit icons added somewhere else in the document by `{% do doc.icons.useIcon %}` - so yes only the clipboard icon.

**Comment by matthiasrohmer** [comment]
@sebastianbenz, could you please review this as I've also contributed? I'll fix the failing teaser text test tomorrow but besides that, this should be finally good to go and would be great to have out the door also as part of #3330.

File: platform/lib/pipeline/componentReferenceImporter.js (hunk 10 above)
Comment (matthiasrohmer) [reply platform/lib/pipeline/componentReferenceImporter.js]: Moved to a configuration file to be aligned with the behaviour of SpecImporter: platform/config/imports/componentReference.json

File: frontend/templates/views/partials/component-intro.j2 [hunk 11]
```diff
@@ -0,0 +1,60 @@
+{% do doc.styles.addCssFile('css/components/molecules/code-snippet.css') %}
+{% do doc.amp_dependencies.add('amp-fx-collection', '0.1') %}
```
Comment (sebastianbenz) [inline frontend/templates/views/partials/component-intro.j2]: why is this needed here?

File: platform/lib/build/samplesBuilder.js [hunk 12]
```diff
@@ -383,13 +386,28 @@ class SamplesBuilder {
...
+
+      await writeFileAsync(COMPONENT_SAMPLES_DEST, JSON.stringify(this._componentSamples), {
+        flag: 'w+',
       });
-      this._log.success('Wrote sample sitemap.');
+
+      this._log.success('Wrote sample sitemap and component samples file.');
     } catch (_) {
-      this._log.info('Samples sitemap already exists');
+      this._log.error('Writing samples builder meta files failed:', e);
```
Comment (sebastianbenz) [inline platform/lib/build/samplesBuilder.js]: I know this was the same before, but wouldn't it be better to fail the build here?

File: platform/lib/build/samplesBuilder.js (hunk 9 above)
Comment (sebastianbenz) [reply platform/lib/build/samplesBuilder.js]: ping

File: platform/lib/build/samplesBuilder.js (hunk 1 above)
Comment (sebastianbenz) [reply platform/lib/build/samplesBuilder.js]: ping

**Comment by sebastianbenz** [comment]
What's the status of this?
